Make TensorIndex track valence and index type

macedo22 opened the pull request:
## Proposed changes

This PR makes `TensorIndex` keep track of valence and whether the index is a spacetime or spatial index.

It also adds some new checks that assert that the valences and index types of the generic indices of the tensor as an expression (the `TensorIndex`s) are the same as the valences and index types of the indices in the tensor's index list. See **Further comments** for more details on the new checks.

A related check that was added was one that asserts that no generic index type is repeated (i.e. can't do `R(ti_i, ti_i)`)

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments
Given this tensor and expression:
```
Tensor<double, Symmetry<2, 1>,
         index_list<SpacetimeIndex<3, UpLo::Up, Frame::Grid>,
                    SpacetimeIndex<3, UpLo::Lo, Frame::Grid>>>
      R{};
auto R_expr = R(ti_A, ti_a);
```
`R`'s first index is an upper spacetime index and its second index is a lower spacetime index. The asserts will make sure that the first index in the expression (`ti_A`) is also upper and spacetime, and that the second index in the expression (`ti_a`) is also lower and spacetime.

fmahebert commented:
👍  I'll take a look towards the middle/end of next week!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -40,16 +40,12 @@ template <typename ReplacedArg1, typename ReplacedArg2, typename T, typename X,
-    tmpl::erase<tmpl::erase<Args, tmpl::index_of<
-                                      Args, TensorIndex<ReplacedArg2::value>>>,
-                tmpl::index_of<Args, TensorIndex<ReplacedArg1::value>>>>::type;
+    tmpl::erase<tmpl::erase<Symm, tmpl::index_of<Args, ReplacedArg2>>,
```
This would probably be a lot easier to read if you defined type aliases `contracted_symm`, `contracted_index_list` and `contracted_args` above

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -185,11 +185,32 @@ namespace detail {
 SPECTRE_ALWAYS_INLINE static constexpr auto fully_contract(const T& t) {
+  // TensorIndex types of pair of indices to contract
+  using lower_tensorindex = ti_contracted_t<I, UpLo::Lo>;
+  using upper_tensorindex = TensorIndex<lower_tensorindex::value + 1, UpLo::Up>;
```
why the `+1` here? I don't understand that, sorry. This seems inconsistent with using the same values in `TensorExpression.hpp`

macedo22 replied:
I think here it wasn't a bug, but just the convention that was originally in place when I first started working on the contractions (I can't tell which commit this is a comment on from the PR's main page). I believe the convention at this point was to have the first pair of contracted indices (the "replaced args") be `TensorIndex<1000>` and `TensorIndex<1001>`, and if there was another pair, they were `TensorIndex<1002>` and `TensorIndex<1003>`. See the example I gave on the previous PR [here](https://github.com/sxs-collaboration/spectre/pull/2572#discussion_r519142294).

So I believe this `+1` was just a reflection of this convention, not a bug. However, I updated the convention to now be `TensorIndex<1000>` and `TensorIndex<1001>` for the first pair, then `TensorIndex<2000>` and `TensorIndex<2001>` for the next pair as of d6a80bf. This also generalizes for cases where we have > 2 contractions if need be (looking at you, inner products).

Another important note is that I had thought the original intention (when I first started working on the contraction code) was to use this convention not only so that there is a predictable/easy way to find the contracted indices, but also to distinguish the upper and lower indices in the pair - meaning, it seemed intentional that they were given different values, as opposed to both being, for example `TensorIndex<1000>`. I guess now that valence is tracked, now there is less of a need for that, because you can distinguish them through their valences. However, with my updates, one nice thing is that the convention for lower contracted indices is `TensorIndex<1000>`,  `TensorIndex<2000>`, `TensorIndex<3000>`, etc. and upper indices is `TensorIndex<1001>`,  `TensorIndex<2001>`, `TensorIndex<3001>`. So, valence is encoded in the last digit. However, again, I'm not sure what real benefit this has now that we have valence to distinguish between an upper and lower index pair that needs to be contracted (before , upper and lower would have been indistinguishable via type).

Lastly, I also liked the idea of the pairs having different values simply because `evaluate` currently assumes that the list of indices for the RHS and the LHS (separately) are unique. Currently, `LhsSymmAndIndices` in `Evaluate.hpp` does the reordering of the symmetry and indices based on this assumption. Of course, that isn't a reason to keep the convention, but it is a reason why I shamelessly liked keeping the convention while updating this code.

If you think it's better that the upper and lower indices have the same values, let me know! I'll think a bit more on this, too.

nilsdeppe replied:
Okay, thanks for the detailed explanation! I think since the valence is now tracked it would be nice to have them have the same value unless that breaks some code. So feel free to delay that until you have products working :) 

I like that generalized to use 2000, 3000, etc :)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -187,7 +187,7 @@ template <template <typename> class TE, typename ReplacedArgList, size_t I,
   // TensorIndex types of pair of indices to contract
   using lower_tensorindex = ti_contracted_t<I, UpLo::Lo>;
-  using upper_tensorindex = TensorIndex<lower_tensorindex::value + 1, UpLo::Up>;
+  using upper_tensorindex = ti_contracted_t<I, UpLo::Up>;
```
aha, now you found the bug :)

macedo22 replied:
See [the above comment](https://github.com/sxs-collaboration/spectre/pull/2584#discussion_r521122563)

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -265,6 +265,10 @@ class Tensor<X, Symm, IndexList<Indices...>> {
             Requires<std::conjunction_v<tt::is_tensor_index<N>...>> = nullptr>
   SPECTRE_ALWAYS_INLINE constexpr auto operator()(
       N... /*meta*/) const noexcept {
+    static_assert(tmpl::is_set<N...>::value,
```
Maybe we should rename `N` to `TensorIndices` or `TensorIndicesInExpression`? I think that would be a lot easier to read (the bad name is my fault, I know...)

macedo22 replied:
Yah, I agree with this change!

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -269,6 +269,10 @@ class Tensor<X, Symm, IndexList<Indices...>> {
                   "Cannot create a tensor expression with a repeated generic "
                   "index. If you intend to contract, ensure that the indices "
                   "to contract have opposite valences.");
+    static_assert(std::is_same_v<tmpl::integral_list<UpLo, N::valence...>,
```
Nice! I forgot about checking this :P

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -31,12 +31,13 @@
  * type `TensorIndex<0, UpLo::Lo>` and `TensorIndex<1, UpLo::Up>` respectively.
  */
-template <std::size_t I, UpLo Valence>
+template <std::size_t I, UpLo Valence, bool IsSpacetime = I < 8>
```
I think the compiler will scream unless you put `(I < 8)`. I can't remember when it gets angry, but it doesn't like angle brackets in templates.

What if we make the spatial indices start at 100? Is there any thing that breaks with that? This would allow adding e.g. `a0, a1, a2, a3` in the future when someone decides they need a rank 50 tensor...

macedo22 replied:
Hmm I think this comment must be on an outdated commit. The compiler didn't complain about this, but doxygen definitely gave a warning. This was fixed in deb8517. Good eye!

I think that changing the spatial index values to start at 100 doesn't break anything. The only constraints to be aware of in the current code, is that contracted indices start at 1000. Anyone looking for a rank 900 spatial tensor? :)

I'll change this!

nilsdeppe replied:
Yea, it was outdated, you're right.

Haha if someone wants a 900 rank tensor they can rewrite the code :P

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -273,6 +273,17 @@ class Tensor<X, Symm, IndexList<Indices...>> {
                   "The valences of the generic indices in the expression do "
                   "not match the valences of the indices in the Tensor.");
+    static_assert(
+        std::is_same_v<
```
I think this:
```cpp
(... and (N::is_spacetime and Indices::index_type == IndexType::Spacetime))
```
https://en.cppreference.com/w/cpp/language/fold

macedo22 replied:
Wow, that syntax (and language feature) is really interesting. I'll try this!

macedo22 commented:
@fmahebert Thanks so much for offering to take a look at this sometime this week! I wanted to let you know that I just spoke with Nils, and he presented a different implementation for a portion of this contraction code. My plan is to first address some of his change requests, squash, and _then_ update the code with a portion of it being implemented differently.

I wanted to notify you of this, because I wouldn't want you to have to review twice! Of course, I would accept any feedback beforehand, but just know portions may change a decent bit. I'll ping you once I've done this to let you know that it's re-ready

fmahebert commented:
@macedo22 Okay, sounds good — thanks for letting me know! I'll wait for you to ping me 🙂

nilsdeppe reviewed (commented):
Looks good to me! Please go ahead and squash :)

I'm not sure how many commits. Probably one for most and then some of the renames into separate commits to make it easier to review?

macedo22 commented:
@nilsdeppe I added in a commit to change starting the spatial indices at 100 and then squashed pretty conservatively just to err on the side of clarity for review.

I think everything thus far has been addressed, but I'll ping you once I've reimplmented the recursive contractions

nilsdeppe commented:
All looks good! Feel free to ping Francois if you are ready for him to take a look :)

macedo22 commented:
@nilsdeppe (After the small fix in e6a4f78) I went ahead and moved the `Tensor::operator()` logic into `contract` in e96881a. I think it may have required a little bit more changes than may have been anticipated. In addition, 4e00be1 moves the logic that determines which contracted index comes first (smaller position of the multi-index) from `contract` into `TensorContract`, because I think it's more robust for it to be determined in `TensorContract` instead of assuming that `Index1` in `TensorContract` is the first one. Because e96881a involved quite a bit of changes (maybe enough that it deserves its own PR) and because if e96881a looks good to you I'm not sure if 4e00be1 is better, I wanted to ask if you could review the changes before asking for Francois' review.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -175,56 +175,50 @@ struct TensorContract
+        t) noexcept {
+  if constexpr (tmpl::size<RepeatedTensorIndexValueList>::value == 0) {
+    // There aren't any repeated indices, so we just return the input
+    return ~t;
```
why ~t instead of `t`? I'm not saying this is incorrect, just trying to understand :)

macedo22 replied:
This was discussed outside of GitHub, but summarizing here for documenting the reason:

Let's say we have a rank 2 `Tensor` named `R` and the expression, `R(ti_I, ti_i)`. When `contract` is called for the first time, the `else` case will be entered, which calls `contract` again and passes to it a `TensorContract` object. Then, the `if` case is entered in this second call. `t` is a `TensorContract` here, but without the `~`, the return type will be its parent class, `TensorExpression<TensorContract<.., Tensor, ...>...>`. This is a problem because when we recurse down and call `TensorExpression::get` (multi-index version) to grab values from the original `Tensor` (`R`), we run into an issue where that expression's `t_` refers to itself instead of the original `Tensor`. This is because a `TensorContract` _is_ a `TensorExpression` of itself. The parent class should never be instantiated or used for this reason.

Instead, we need to downcast with `~t` here to ensure that a returned `TensorContract` is interpreted as a `TensorContract` instead of a `TensorExpression` of itself.

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -281,19 +281,14 @@ class Tensor<X, Symm, IndexList<Indices...>> {
-                          repeated_rhs_tensorindex_values>,
-          1, tmpl::size<repeated_rhs_tensorindex_values>::value>(*this);
-    }
+    using tensorindex_value_list =
```
Why can't the logic here for `placed_tensorindex_value_list`, etc. be inside contract? I was initially hoping this function would just be:
```cpp
return contract(TE<tmpl::list<TensorIndices...>>{*this});
```

macedo22 replied:
Nuked in 062e2b3

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -175,56 +175,50 @@ struct TensorContract
+    //
+    // e.g. `R(ti_A, ti_b, ti_a)` :
+    // - `first_replaced_tensorindex` refers to the contracted TensorIndex type
+    // for `ti_A`
```
[nitpick] I find bullet lists in comments are easier to read when continued lines are indented

macedo22 replied:
Agreed! I originally had it this way, so I'm not sure if `git-clang-format` changed this. I'll change it back, regardless

macedo22 replied:
Upon looking at this again, I think this was on an outdated commit, as the indentation looks okay in 4e00be1, which is the most recent commit at this point

macedo22 commented:
@nilsdeppe I removed the whole contracted index value replacement and changed `TensorIndex` to encode its valence and index type within its value. I'm pretty happy with how much things have been simplified by this change.

Not directly related, but I realized when checking the documentation generated, no documentation is generated for `contract`, and the documentation for `TensorIndex` looks pretty wonky. Let me know if you see any syntactical mistakes I may have made there, as I'm not sure how to fix these two issues.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -27,20 +44,34 @@
+       (I >= spatial_sentinel and I < upper_spatial_sentinel))
+          ? UpLo::Lo
+          : UpLo::Up;
+  static constexpr bool is_spacetime = (I < spatial_sentinel) ? true : false;
```
fun fact! `(I < spatial_sentinel) ? true : false` is equal to `(I < spatial_sentinel)`. If you wrote this out explicitly it would be:
```cpp
bool check_true = (I < spatial_sentinel);
if (check_true == true) {
  return true;
} else {
  return false;
}
```

macedo22 replied:
Really not sure why I had this as a ternary expression, whoops. My best guess is that this is a vestige of something that was originally some complex ternary expression. Thanks for catching this!

macedo22 replied:
Fixed!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -17,6 +17,23 @@
+// TensorIndex values to separate upper indices from lower indices and spatial
+// indices from spacetime indices.
+//
+// Spacetime indices are represented by values [0, `spatial_sentinel`) and
```
Maybe add a little bit about why this is being done:
```cpp
// Tensor expressions perform as many calculations as possible in a constexpr context,
// which means working with fundamental types, specifically integer types, is easiest. By
// using sentinel values defined in one location we can easily control the encoding without
// having magic values floating around in many places. Furthermore, encoding all the
// information in the `size_t` means that when a failure occurs in one of the constexpr
// calculations it is reasonably easy to debug because, while encoded, the full type information
// is present. This approach can effectively be thought of as using specific bits in the size_t to
// mark information, using the size_t more as a bitfield than anything else. For human readability
// we use base-10 numbers instead of base-2 values that would truly set individual bits.
```

macedo22 replied:
Done!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -196,27 +196,41 @@ struct TensorContract
+ * one pair of upper and lower indices. If there is more than one pair of
+ * indices to contract, subsequent contraction expressions are recursively
+ * created, nesting one contraction expression inside another.
+ *
```
I'd also explain the design choice (since anyone reading this documentation is trying to understand the design itself):
```cpp
 * \note The `IndicesToContract` are passed in rather than computed internally to avoid recomputing
 * the new list of indices that are contracted inside the function during each recursion. Instead, we
 * only need to do a `tmpl::pop_front`
```
Having said all that, considering we will have at most ~10 indices, that's an easy computation to do and will basically be free, even compared to `pop_front` so I would advocate for the slightly simpler interface at the cost of possibly re-checking a few indices. Additionally, I think you only need to find the _first_ repeated index each time, so you could do it with no extra cost. You could even do this in `constexpr` land, I think, by converting the `TensorIndexList` to a `std::array` and feeding it through
```cpp
constexpr auto tensor_index_list = make_cpp20_array_from_list<TensorIndexList>();
constexpr auto first_repeated_index_value_it = 
  alg::find_if(tensor_index_list, [&tensor_index_list](size_t tensor_index_value) {
  return alg::found(tensor_index_list,  opposite_valence(tensor_index_value));
  });
if constexpr (first_repeated_index_value_it == tensor_index_list.end()) {
  return ~t;
} else {
  // ...
}
```
What do you think?

nilsdeppe replied:
Thinking about this some more, couple for loops might be the best option, actually. Because then you can also return the location of the repeated indices:
```cpp
constexpr std::pair first_repeated_index_locations =
  []() noexcept {
    constexpr auto tensor_index_list = 
      make_cpp20_array_from_list<TensorIndexList>();
    for (size_t i = 0; i < tensor_index_list.size(); ++i) {
      const size_t index_to_find =
            opposite_valence(tensor_index_list[i]);  // NOLINT
      for (size_t j = i + 1; j < tensor_index_list.size(); ++j) {
        if (index_to_find == tensor_index_list[j]) { // NOLINT
          return std::pair{i, j};
        }
      }
    }
    return std::pair{std::numeric_limits<size_t>::max(),
                     std::numeric_limits<size_t>::max()};
  }();
constexpr std::pair no_repeated_sentinel{std::numeric_limits<size_t>::max(),
                                         std::numeric_limits<size_t>::max()};
if constexpr (first_repeated_index_locations == no_repeated_sentinel) {
  return ~t;
} else {
  // ...
}
```

macedo22 replied:
Hmm I think I'm not following what the motivation for this change would be. To me, it seems like more verbose code that's less easy on a reader. Is the motivation just to avoid calling `contract` with any template parameters, or is the motivation that this implementation is faster than using `tmpl::pop_front`? Trying to figure out if you're mainly being motivated by being able to call `contract` without a parameter or a performance benefit.

In general though, I think this should work.

nilsdeppe replied:
The big motivation is removing more template code by replacing it with runtime code. You are replacing the what I would argue is very difficult to read:
```cpp
template <typename Seq, typename S, typename E>
struct contracted_indices_helper {
  using type = typename std::conditional<
      std::is_same<tmpl::count_if<
                       Seq, std::is_same<::tensorindex_with_opposite_valence<E>,
                                         tmpl::_1>>,
                   tmpl::size_t<1>>::value and
          std::is_same<
              tmpl::index_of<S, ::tensorindex_with_opposite_valence<E>>,
              tmpl::no_such_type_>::value,
      tmpl::push_back<S, E>, S>::type;
};

template <typename TensorIndexList>
using contracted_indices =
    tmpl::fold<TensorIndexList, tmpl::list<>,
               detail::contracted_indices_helper<tmpl::pin<TensorIndexList>,
                                                 tmpl::_state, tmpl::_element>>;
```
with the much easier for loops. You could also factor the for loops into a separate helper function. The above TMP is, I think, much harder to understand than two nested for loops.

```cpp
constexpr std::pair<size_t, size_t> first_contracted_indices_locations(
      const std::array<size_t, Rank>& tensor_index_list) noexcept {
    for (size_t i = 0; i < tensor_index_list.size(); ++i) {
      const size_t index_to_find =
            opposite_valence(tensor_index_list[i]);  // NOLINT
      for (size_t j = i + 1; j < tensor_index_list.size(); ++j) {
        if (index_to_find == tensor_index_list[j]) { // NOLINT
          return std::pair{i, j};
        }
      }
    }
    return std::pair{std::numeric_limits<size_t>::max(),
                     std::numeric_limits<size_t>::max()};
}
```
I find is much easier to read than the 12 lines of TMP.

Does that make my motivation clear? Basically, I'd like to replace the `contracted_indices` TMP code with the constexpr function because constexpr code is so much easier to read.

Edit:
It also reduces the number of loops you need to do through the list from ~3 to 1.5, so just in that sense (never mind the faster performance in general of constexpr over TMP) the code would compile faster. I must also say that it took me a bit to follow the logic of the TMP, despite probably having written the original version of it.

macedo22 replied:
Gotcha, this makes sense, and I'm in agreement. I think I was narrowly looking just at the `contract` function, but I agree. This is easier to follow than `contracted_indices`. I'll update this, and thanks for the suggestion!

macedo22 replied:
@nilsdeppe Went ahead and implemented this in 8d86973 and swapped out `contracted_indices`. Thanks again for this suggestion.

nilsdeppe reviewed (commented):
This looks great! A couple thoughts now that I can see it in action that might make the code even easier to read :)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -63,14 +63,29 @@ struct TensorIndex {
-                                       : T::value + upper_sentinel>;
+SPECTRE_ALWAYS_INLINE static constexpr size_t
+get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
+  return ((i >= upper_sentinel and i < spatial_sentinel) or
```
[optional] you no longer need this to be a ternary. and if-else block might be easier to read

macedo22 replied:
Interestingly, when this is converted into `if constexpr` / `else`, the compiler complains about these expressions not being evaluable as `constexpr`:

```
SPECTRE_ALWAYS_INLINE static constexpr size_t
get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
  if constexpr ((i >= upper_sentinel and i < spatial_sentinel) or
                (i >= upper_spatial_sentinel)) {
    // `i` represents an upper index, so return the lower index's encoding
    return i - upper_sentinel;
  } else {
    // `i` represents a lower index, so return the upper index's encoding
    return i + upper_sentinel;
  }
}
```

```
SPECTRE_ALWAYS_INLINE static constexpr size_t
get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
  constexpr bool is_upper = ((i >= upper_sentinel and i < spatial_sentinel) or
                             (i >= upper_spatial_sentinel));
  if constexpr (is_upper) {
    // `i` represents an upper index, so return the lower index's encoding
    return i - upper_sentinel;
  } else {
    // `i` represents a lower index, so return the upper index's encoding
    return i + upper_sentinel;
  }
}
```

It looks to me like if I want to turn this ternary into an `if constexpr`/`else`, then the function will need to be templated on the value instead of it being passed in as a normal argument. The `if constexpr`/`else` might be somewhat clearer, but then I also don't need to know if we need to create so many instantiations of the function template if it works as is.

Let me know your thoughts. Maybe I'm misunderstanding what you were suggesting, here!

macedo22 replied:
It is now an if-else!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -78,29 +83,49 @@ struct TensorContract
+          : replaced_arg1_position;
+
+  // first and second \ref SpacetimeIndex "TensorIndexType"s to contract
+  using CI1 = tmpl::at_c<IndexList, first_position_to_contract>;
```
what about naming these `contracted_upper_index` and `contracted_lower_index`?

macedo22 replied:
I was on the fence about doing this here in this PR because there are a number of aliases I want to rename, but I'll do this one now!

macedo22 replied:
Done

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -168,29 +196,88 @@ struct TensorContract
+      if (opposite_value_to_find == gsl::at(tensorindex_values, j)) {
+        // We found both the lower and upper version of a generic index in the
+        // list of generic indices, so we return this pair
+        return std::pair{current_value, opposite_value_to_find};
```
[optional] what if you did the check here to return `upper, lower`? Then you don't need the check in the Contract struct.

macedo22 replied:
Done! (Except, instead of the upper valence index's position and lower valence index's position, I return the smaller position and larger position, regardless of valence)

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -168,29 +196,88 @@ struct TensorContract
+        TensorIndex<first_tensorindex_values_to_contract.first>;
+    using opposite_tensorindex_to_contract =
+        TensorIndex<first_tensorindex_values_to_contract.second>;
+    return contract(TensorContract<tensorindex_to_contract,
```
[optional] I'm now wondering if it makes more sense to pass in the locations of the indices to contract, rather the indices them selves. Combined with guaranteeing `upper, lower` in the ordering I think it also means you don't need to do the `tmpl::index_of` call inside the `TensorContract` struct. It might then be good to move `TensorContract` into the `detail` namespace, making this function the interface to it. Thoughts?

macedo22 replied:
I don't have a strong preference on this. I think I originally had it this way with the `TensorIndex`s instead of their positions just so that `TensorContract` could work out details internally in one place instead of it being externally dictated in multiple. If there were multiple spots where `TensorContract` would be called, then we might have to work out the index positions each time, which seemed like more work to use `TensorContract` as an interface than just giving it the two `TensorIndex`s. It was like it kind of felt more fragile for something external to be telling `TensorContract` the positions of the two indices instead of `TensorContract` working it out for itself, especially with both a contracted args list and uncontracted args list as things to deal with. However, since it seems `TensorContract` shouldn't really be used anywhere else, I don't think this point matters so much. I'll make this change!

Regarding putting `TensorContract` in the `detail` namespace, I think I don't naturally see it there only because it's one of the three expression types making up this tensor expression library/interface. So it doesn't really feel like background details, as a whole. Is there a property about it that makes you feel it belongs in the `detail` namespace, whereas `AddSub` and `Product` wouldn't? I think I'm not as familiar with when a whole class/struct makes more sense to stick in `detail`.

nilsdeppe replied:
That all sounds good. Let's not move it into a detail namespace. I agree consistency with `AddSub` and `Product` will be good!

macedo22 replied:
Done! (Except, instead of the upper valence index's position and lower valence index's position, I return the smaller position and lower position, regardless of valence)

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -261,27 +261,27 @@ class Tensor<X, Symm, IndexList<Indices...>> {
-        tmpl::size<repeated_args_list>>(*this);
+  template <
+      typename... TensorIndices,
+      Requires<std::conjunction_v<tt::is_tensor_index<TensorIndices>...>> =
```
I think this `Requires` argument can be a `static_assert` it's mostly used for error message info, not for SFINAE and overload resolution.

macedo22 replied:
Done!

nilsdeppe reviewed (commented):
All looks great! I think we're pretty much there a few optional ideas for you to think about. They might be bad so just decline them if you think they're not good :) 

I'm happy for you to squash everything

macedo22 commented:
@nilsdeppe I think I've addressed everything, but let me know if the implementation changes differ from your expectations

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -93,10 +93,14 @@ struct TensorIndex {
  */
 SPECTRE_ALWAYS_INLINE static constexpr size_t
 get_tensorindex_value_with_opposite_valence(const size_t& i) noexcept {
```
when you squash: you don't need to take types like `size_t` and `double` by reference because references are effectively pointers, pointers are 8 bytes in size, just like `size_t` and `double`, so it's actually more efficient to "copy" the value than to create a pointer. In practice, since this is always inlined it probably won't matter, but a useful FYI :)

macedo22 replied:
Good point regarding the equivalent number of bytes. I'll update it regardless of the inlining

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -261,27 +261,28 @@ class Tensor<X, Symm, IndexList<Indices...>> {
+  template <typename... TensorIndices>
+  SPECTRE_ALWAYS_INLINE constexpr auto operator()(
+      TensorIndices... /*meta*/) const noexcept {
+    static_assert(std::conjunction_v<tt::is_tensor_index<TensorIndices>...>,
```
It's more efficient (in general) to use a fold expression rather than `conjunction`. The reason being is that `conjunction` can generate a bunch of different types, while folds done. E.g. `... and tt::is_tensor_index_V<TensorIndices>`

nilsdeppe reviewed (commented):
All looks great! Please go ahead and squash things, including the last minor comments

macedo22 commented:
@nilsdeppe Thanks for all of your feedback. I went ahead and put in the last couple changes you requested. When going to squash, I did it very conservatively because it might make it easier for a second reviewer to see the changes, and also because I didn't see very many places where there was a very natural grouping of consecutive commits. Let me know if you'd prefer for me to squash all of this down into a commit or two now, as opposed to at the end, when merging.

macedo22 commented:
@fmahebert Whenever you get the chance, this should be ready for review. A big change that was added to this PR on top of incorporating valence and index type into `TensorIndex` (including changing the encoding of the values) was significantly changing how contractions are recursively done. One big piece to this is that instead of `Tensor::operator()` determining if a contraction needs to happen, all expressions are passed to `contract` in `Contract.hpp`, and the determination is done there. There's a bit more documentation for the functions this time around, but please let me know if you have questions! The sort of middle level details of how contraction is carried out changed somewhat significantly in this PR.

fmahebert commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -25,7 +54,7 @@
  * type of expressions to work:
  * \code{.cpp}
- * auto T = evaluate<ti_a_t, ti_b_t>(F(ti_a, ti_b) + S(ti_b, ti_a));
+ * auto T = evaluate<ti_a_t, ti_B_t>(F(ti_a, ti_B) + S(ti_B, ti_a));
```
For my understanding: is this just a pedagogical change, i.e., to illustrate mixing of upper/lower indices? Or was this malformed? It seems like it should be fine..

Now the (potential) change request: in the lines below you have a `ti_b` that should maybe be updated to `ti_B` ?

macedo22 replied:
Good eye. I did intentionally change this when I changed `TensorIndex` to be templated on valence, because I was showing `TensorIndex<0, UpLo::Lo>` for `ti_a` and `TensorIndex<1, UpLo::Up>` for `ti_B`. However, I didn't change this back after the rework for the `TensorIndex` encodings with the sentinels, where it makes more sense to show the example with `ti_b` over `ti_B`, because `TensorIndex<1>` is simpler than `TensorIndex<upper_sentinel + 1>` for demonstrating the point that they're placeholders. I've changed this back to use `ti_b`. Thanks!

fmahebert commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -36,18 +65,43 @@ struct TensorIndex {
+  if ((i >= upper_sentinel and i < spatial_sentinel) or
+      (i >= upper_spatial_sentinel)) {
+    // `i` represents an upper index, so return the lower index's encoding
+    return i - upper_sentinel;
```
There's an edge-case bug here: if the input to this function is, e.g., i = 812, the function will return 712 which is still an upper index. Fundamentally, this arises because the interval for the lower indices = [0, upper_sentinel) is smaller than the interval for the upper indices = [upper_sentinel, spatial_sentinel), so there are plenty of upper indices that have no matching lower index.

I don't think this is a deep problem, because the sentinels you've chosen should provide more than enough indices for our needs, but it might be good (thinking to future bug avoidance) to check the inputs at least fall within the "reasonable" ranges? An assertion like `(i < 2*upper_sentinel or (i >= spatial_sentinel and i < spatial_sentinel + 2*upper_sentinel))` at the top of the function may be a sufficient fix. Or perhaps you have other ideas?

macedo22 replied:
Good catch. I've gone ahead and changed the convention so that `upper_sentinel` and `upper_spatial_sentinel` are half-way points, and also added a `max_sentinel` to denote the maximum value. So now:
- lower spacetime: [0, 500)
- upper spacetime: [500, 1000)
- lower spatial: [1000, 1500)
- upper spatial: [1500, 2000)
- max: 2000

I also added an assert in `get_tensorindex_value_with_opposite_valence` and a small accompanying unit test to address both regular and edge cases. The unit test is predicated on the ordering of the four categories of indices above (e.g. values immediately above the upper spacetime values are lower spatial values), but because this is such a small test and I don't see this convention being updated often (if at all), I thought it was okay for the test to be predicated on the assumption of the orderings. If not, let me know!

macedo22 replied:
The clang-tidy warning about the `assert` I added in `get_tensorindex_value_with_opposite_valence` reads strangely to me, but I'm thinking it's because I'm making a runtime assertion in a `constexpr` context. I wasn't able to use `ASSERT` since the second argument is a string and using a string (a non-literal) doesn't work because the function is `constexpr`, and it needs to be `constexpr` because `contract` depends on that. I'm thinking that the solution is just to make `get_tensorindex_value_with_opposite_valence` be templated on the input value `i` instead of it being a regular argument. What do you think? Let me know if I'm misinterpreting the clang-tidy warning!

fmahebert commented on src/DataStructures/Tensor/Expressions/Contract.hpp:
```diff
@@ -168,29 +189,85 @@ struct TensorContract
+ * any indices to contract
+ *
+ * \details If there are no indices to contract, the input TensorExpression is
+ * simply returned. If not, a contraction expression is created for contracting
```
[nitpick] "If not" -> "Otherwise" (because the prior sentence describes the negative case, it's confusing to start this one with "if not")

macedo22 replied:
Done!

fmahebert reviewed (commented):
This looks like a really nice change! Only a few small comments from me.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/TensorExpression.hpp:
```diff
@@ -31,23 +62,49 @@
+ */
+SPECTRE_ALWAYS_INLINE static constexpr size_t
+get_tensorindex_value_with_opposite_valence(const size_t i) noexcept {
+  assert(i < max_sentinel);
```
Can you just add `  // NOLINT` to the end of this line to disable clang-tidy? As far as I can tell it's complaining about code in the standard library and we can't exactly change that... Feel free to squash that in immediately

nilsdeppe replied:
So to explain a bit on why the code is actually "okay" and an alternative that is more visibly okay we need to know some rules... C++ always constexpr evaluation as long as the _call path_ is constexpr. That is, you can have

```cpp
constexpr size_t blah(const size_t i) {
  if (i >= 10) {
    std::cout << "Yoohoo! Big summer blowout!\n";
  }
  return i;
}
```
and if `i < 10` this can be evaluated constexpr, otherwise you get the following "fun":
```shell
test.cpp:67:28:   in ‘constexpr’ expansion of ‘blah(10)’
test.cpp:59:18: error: call to non-‘constexpr’ function ‘std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const char*) [with _Traits = std::char_traits<char>]’
   59 |     std::cout << "Yoohoo! Big summer blowout!\n";
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/10/iostream:39,
                 from test.cpp:6:
/usr/include/c++/10/ostream:606:5: note: ‘std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const char*) [with _Traits = std::char_traits<char>]’ declared here
  606 |     operator<<(basic_ostream<char, _Traits>& __out, const char* __s)
      |     ^~~~~~~~
```
So basically, inside `assert` we have code like:
```cpp
if (not (ARGUMENT)) {
  printf("Blah blah blah");
  std::abort();
}
```
The non-constexpr code is not evaluated _unless_ `ARGUMENT` is `false`, which causes the `assert` to fail. This, of course, relies on the implementation-defined internals of `assert` (though I'm not sure how exactly you could have an implementation that's sufficiently weird that it wouldn't behave this way). A safer way to do the error handling would be:
```cpp
constexpr size_t blah(const size_t i) {
  if (i >= 10) {
    throw std::domain_error{"Value i must be less than 10" + std::to_string(i)};
  }
  return i;
}
```
The disadvantage compared to the runtime error always being you don't actually get to know the value of `i` but I'm unaware of a way to fix that.

fmahebert replied:
Thanks for the example @nilsdeppe!

@macedo22 I think I agree with Nils's suggestion on NOLINT'ing the assert. I don't think it makes sense to increase the complexity of the code here by adding templates, for instance.

macedo22 replied:
Thanks for the example, Nils! I also just played around with this locally for a few different scenarios. Makes sense. I've gone ahead and just suppressed the linting warning

macedo22 commented:
@nilsdeppe @fmahebert Let me know if you would prefer this all squashed this into one commit, multiple, or reorganized some other way

nilsdeppe reviewed (commented):
To me this seems like roughly one commit would be right. If it's easy to have a separate commit (or more than one) for some of the renames that would be good, but don't put a lot of work into it. Remember that we need each commit merged to compile and pass tests :)

fmahebert commented:
I'll just +1 what Nils said.

macedo22 committed:
Make TensorIndex track valence and index type

macedo22 committed:
Rename Tensor::operator() generic indices tparam

macedo22 committed:
Rename ContractedType generic indices tparam

macedo22 committed:
Rename TensorContract IndexType aliases

macedo22 commented:
Rebased on develop and reorganized to include clear commits for renames :+1:
